### PR TITLE
Refactor flake8 handling and tool compatibility

### PR DESCRIPTION
**Pull request by trws**
This PR does three related things to try to improve developer tooling quality of life:

1. Adds new options to `.flake8` so it applies the rules of both `.flake8` and `.flake_package` based on paths in the repository.
2. Adds a re-factoring of the `spack flake8` logic into a flake8 plugin so using flake8 directly, or through editor or language server integration, only reports errors that `spack flake8` would.
3. Allows star import of `spack.pkgkit` in packages, since this is now the thing that needs to be imported for completion to work correctly in package files, it's nice to be able to do that.

I'm sorely tempted to sed over the whole repository and put `from spack.pkgkit import *` in every package, but at least being allowed to do it on a per-package basis helps.

As an example of what the result of this is:

```
~/Workspace/Projects/spack/spack develop* ⇣
❯ flake8 --format=pylint ./var/spack/repos/builtin/packages/kripke/package.py
./var/spack/repos/builtin/packages/kripke/package.py:6: [F403] 'from spack.pkgkit import *' used; unable to detect undefined names
./var/spack/repos/builtin/packages/kripke/package.py:25: [E501] line too long (88 > 79 characters)

~/Workspace/Projects/spack/spack refactor-flake8*
1 ❯ flake8 --format=spack ./var/spack/repos/builtin/packages/kripke/package.py

~/Workspace/Projects/spack/spack refactor-flake8*
❯ flake8 ./var/spack/repos/builtin/packages/kripke/package.py
```

### CODE REVIEW

**Comment by adamjstewart** [comment]
My only concern is that now there is a lot of duplication between `lib/spack/spack/cmd/flake8.py` and `share/spack/qa/flake8_formatter.py`. Is it possible to simplify `spack flake8` to use the formatter you added? If we do, then `spack flake8` will be identical to `flake8` except that it detects which files have changed. That would allow us to test it as well. If we're adding Python files to `share/spack/qa`, I think we might need to explicitly enable coverage on them.

**Comment by trws** [comment]
Absolutely.  I actually started to do it but decided to put this up this way first in case we wanted to preserve that for some reason.  The only reason I can think of we would want to do that is that the plugin system is *relatively* newish to flake8, but since CI runs a new version, and it's only a pip install away I'm not sure that matters so much.  I can refactor the command on top of this shortly.

As to the placement, I put it there just to keep it with the qa stuff, but we can pretty easily move it elsewhere.

**Comment by adamjstewart** [comment]
How new is _newish_? Regardless, I'm still in favor of it for the same reasons you outlined. The only potential problem/annoyance is if we introduce a change that means `spack flake8` doesn't work with Python 2.6 anymore.

**Comment by adamjstewart** [comment]
> CI runs a new version

`spack flake8` CI runs a new version, but we do have unit tests in `lib/spack/spack/test/cmd/flake8.py` that still need to work.

**Comment by trws** [comment]
Support for this kind of local, uninstalled, plugin was added in 3.5.0 which was apparently released in October of 2017.

How do the spack flake8 tests currently work, since flake8 doesn't support python 2.6?

**Comment by trws** [comment]
Rebased.

**Comment by adamjstewart** [comment]
> How do the spack flake8 tests currently work, since flake8 doesn't support python 2.6?

It looks like we use a `skipif` for Python 2.6, so as long as the latest version of flake8 support Python 2.7 then we're good.

**Comment by trws** [comment]
Cool, it claims to so I'll put together that refactor and see what the CI thinks of it.

**Comment by trws** [comment]
That was a little nastier than I thought because of `F811`.  I think I have a reliable workaround for the issue in there, but it's a little bit gnarly.

File: .flake8 [hunk 1]
```diff
@@ -30,3 +30,36 @@
...
 max-line-length = 79
+
+# F4: Import
+# - F405: `name` may be undefined, or undefined from star imports: `module`
+#
+# F8: Name
+# - F821: undefined name `name`
+#
+per-file-ignores =
+  var/spack/repos/*/package.py:F405,F821
```
Comment (tgamblin) [inline .flake8:41]: this is awesome

**Review by tgamblin — APPROVED** [review]

**Pull request merged**